[FLINK-21354][statebackend] Introduce ChangelogStateBackend to delegate state access

curcur opened the pull request:
## What is the purpose of the change

This change is to wrap the existing prod state backend (Rocksdb, Fs, Mem) to delegate state access for these state backends. In the future, we can forward state changes to StateChangeLog when states are updated.

In this PR, we only support keyed-state access.

## Brief changelog

1. Introduce `DelegatingStateBackend` interface for state backend delegation (in the flink-runtime module)
2. Introduce `ChangelogStateBackend` and related delegating states for state delegation (in flink-state-backends module)
3. Implement `ChangelogStateBackend`'s Loader in `StateBackendLoader`.

## Couple of things to highlight in a bit more detail
1. `ChangelogStateBackend` can be enabled through configuration key "state.backend.enable-statechangelog" or through application API directly.
2. `ChangelogStateBackend` Loader encapsulates wrapping logic to `loadStateBackendFromConfig` and `fromApplicationOrConfigOrDefault`. The original loading methods are renamed to `loadUnwrappedStateBackendFromConfig` and `loadFromApplicationOrConfigOrDefaultInternal` respectively, and kept unchanged.
3. Recursive wrapping is not supported:   
        `evn.setStateBackend(new new ChangelogStateBackend(new ChangelogStateBackend(...)))` NOT SUPPORT
4. In the current state backend, state access from applications is interfaced by `InternalKvState`, and `InternalKvState` is where state access is delegated.
5. How a state is physically represented/stored depends on the type of underlying state backend, and we do not need to worry about this part in delegation. Hence they are untouched in this PR.
6. TTL States wrap on top of `InternalKvState` as well. If TTL is enabled, Changelog States wrap on top of TTL States.

## Verifying this change
1. Unit tests
2. We have enabled State Backend Delegation by default to run-through all tests (including end-2-end nightly builds). In the current PR, StateChangeLog is not enabled by default.
3. We will have a follow-up ticket to randomize the tests.  

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (not yet)

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -242,6 +265,53 @@ public static StateBackend fromApplicationOrConfigOrDefault(
+                                .asSubclass(DelegateStateBackend.class)
+                                .getConstructor(DelegatedStateBackend.class);
+
+                Class.forName(ROCKSDB_STATE_BACKEND_FACTORY, false, classLoader);
```
Could you add a comment why is it needed?

curcur replied:
I've added it to test classloader, it should be removed. Sorry.

curcur replied:
Thanks for the careful review, it may cause bugs that are strange/difficult to debug.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/state/ChangelogAggregatingState.java:
```diff
@@ -0,0 +1,118 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.state;
```
`package org.apache.flink.state.changelog;`
?

curcur replied:
I do not mind, either way. Let's maybe decide this after your API changes are finalized as well?

rkhachatryan replied:
They are now merged.

curcur replied:
I put state after changelog because I think these are state-related part of changelog

There may be other parts of changelog that may not be state related, for example writer/client, e.t.c

What is your reason to suggest to move the way around?

rkhachatryan replied:
I think that changelog is only used by backend. Theoretically, backend can use some other mechanism and changelog can be used by other parts.

For example, changelog is not mandatory to wrap incremental heap backend (only the materialization part is required).

curcur replied:
Where ChangelogWriter/client put (currently, in-memory is put in runtime), but what Fs-based is put?
Are they going to be put in different module than changelog state backend?

rkhachatryan replied:
Yes, it will be a different module.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:
```diff
@@ -162,14 +163,19 @@ public static CheckpointStorage load(
         Preconditions.checkNotNull(configuredStateBackend, "statebackend");
 
-        if (configuredStateBackend instanceof CheckpointStorage) {
+        StateBackend rootStateBackend =
+                (configuredStateBackend instanceof DelegateStateBackend)
+                        ? ((DelegateStateBackend) configuredStateBackend).getDelegatedStateBackend()
+                        : configuredStateBackend;
```
DelegateStateBackend (Abstract class) is added exactly for that purpose. 
New delegating/proxying only needs to extends DelegateStateBackend, like ChangelogStateBackend.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -204,6 +222,11 @@ public static StateBackend fromApplicationOrConfigOrDefault(
 
         // (1) the application defined state backend has precedence
         if (fromApplication != null) {
+
+            checkArgument(
+                    !(fromApplication instanceof DelegateStateBackend),
+                    "DelegateStateBackend can not be delegated!");
```
Why not?

rkhachatryan replied:
> Why not?

> Another concern: can we get DelegateStateBackend here after deserialization?

curcur replied:
When I said why not, I was thinking of why does `deserialization` matter here? Or put it another way, `deserialization` in where for what? And why does `deserialization` be a concern?

If you mean the read config deserialization in JM or TM, it does not matter because we do not allow `changelogstatebackend` from the application. Even if we do, that should not be a problem as well, because DelegateStateBackend is a StateBackend which is `java.io.Serializable`.

Or do you mean something else?

rkhachatryan replied:
>  we do not allow changelogstatebackend from the application.

Do you mean user can NOT use java API, only config? Why?
I think user SHOULD be able to configure backend via both API and config.

I mean a scenario when
1. `ChangelogStateBackend` is created in the user app 
2. then stored (serialized) on JM 
3. then deserialized on TM 
4. the passed to this method

P.S.: this method anyways should not perform the check as it is not aware of **why** this check should be done.
[This](https://github.com/apache/flink/pull/14943#discussion_r583600890) is I think the right place.

curcur replied:
But does not ALL OF US agree that `ChangelogStateBackend` is only configured through "flag"? That long discussion with Seth, Piotr, Arvid and you??

Please check the COUPLE OF HILIGHTS of this PR
"ChangelogStateBackend is enabled through configuration key "state.backend.enable-statechangelog". It is not directly exposed to users, because ChangelogStateBackend is an implementation detail and users should NOT be aware of it. In other words, application initialization is not supported:
evn.setStateBackend(new ChangelogStateBackend(...)) NOT SUPPORT"

And that's why I added the check 

"""
if (config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG)) {
            LOG.info(
                    "Delegate State Backend is used, and the root State Backend is {}",
                    backend.getClass().getSimpleName());
"""

Even if it allows to configure through the app, I dont think deserialization would be a problem, which is not now. That's why I ask where the concern comes from.

rkhachatryan replied:
We agreed to use "flag" in **both** config and API. I didn't see this API change in this PR and assumed it's addition is postponed and for now explicit configuration will be used in Flink internally.
I think that for internal use explicit configuration will be needed anyways; and to implement the "flag" in the API we'll have to remove this check.

Furthermore, I don't see any reason why should we disallow users (or us) explicit configuration. If someone has already constructed and passed the backend here then it means it's already exposed.

> I dont think deserialization would be a problem. That's why I ask where the concern comes from.

Please see my [reply above](https://github.com/apache/flink/pull/14943#discussion_r583867401).

curcur replied:
What do you mean by "flag" in API?

rkhachatryan replied:
From the offline discussion you mentioned:
`env.getCheckpointConfig().enableChangelogPersistence();`

> And does the "flag" in config mean the same thing as what is supported in this PR?

I'm not sure if I understand your question. I think this PR doesn't support this  currently; but it should, or at least should allow explicit configuration via API.

curcur replied:
Let's maybe discuss it offline? I think we have some misunderstandings about what is "flag" means. Or we refer to different things. I guess I kind of understanding, but just in case.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/delegate/DelegatedStateBackend.java:
```diff
@@ -0,0 +1,29 @@
+ * An interface for DelegatedStateBackend. A state backend to be delegated must implement this
+ * interface.
+ */
+public interface DelegatedStateBackend extends StateBackend {}
```
1. I am saying `unwrap` is not good, but instead, I feel unwrap is not a concept belonging to `StateBackend `, instead it is a concept belonging to DelegateStateBackend`
2. Currently, you can think of getDelegatedStateBackend in DelegateStateBackend a simplified version of unwrapping.

1. For the delegatee/delegated matching, I do not see how adding unwrap in `StateBackend` can be simpler than the current approach; and there are several different ways we can solve the first problem: for example, the match check can be done when loading explicit delegatee; Or we can put the matching logic in DelegateStateBackend.

> If we add some new delegatee then it's impossible to distinguish for which delegatee given backend can be delegated. For example, we add a new delegating backend with extended logging. How RocksDbBackend can signal that it can NOT work with it? 


2. Yes, it should. Not every state backend is delegable. When implementing a new state backend, it should register the information somewhere whether it is delegable and who can delegate it. Such logic should be put within DelegatedStateBackend, and DelegateStateBackend should use such information to double check. That's the intension to introduce Interface DelegatedStateBackend and Abstract Class DelegateStateBackend

> Every new backend must be marked explicitly

rkhachatryan replied:
The interface is removed as agreed offline.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/delegate/DelegateStateBackend.java:
```diff
@@ -0,0 +1,43 @@
+ * {@code MemoryStateBackend}, {code FsStateBackend} and {@code RocksDBStateBackend}.
+ */
+@Internal
+public abstract class DelegateStateBackend implements StateBackend, java.io.Serializable {
```
Explained above already.

curcur commented:
rebase

curcur commented:
rebase

curcur commented:
solve conflicts with

rkhachatryan commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StateConfigUtil.java:
```diff
@@ -46,19 +46,24 @@ public static StateTtlConfig createTtlConfig(long retentionTime) {
         boolean isRocksDbState =
-                ROCKSDB_KEYED_STATE_BACKEND.equals(stateBackend.getClass().getCanonicalName());
-        boolean isHeapTimer = false;
-        if (stateBackend instanceof AbstractKeyedStateBackend) {
-            // currently, requiresLegacySynchronousTimerSnapshots()
-            // indicates the underlying uses heap-bsased timer
-            isHeapTimer =
-                    ((AbstractKeyedStateBackend<?>) stateBackend)
-                            .requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT);
-        }
+                ROCKSDB_KEYED_STATE_BACKEND.equals(
+                        rootKeyedStateBackend.getClass().getCanonicalName());
```
I think this `getDelegatedKeyedStateBackend` is not necessary anymore as well as pre-existing class check (is rocks db) after you added `requiresLegacySynchronousTimerSnapshots` to `KeyedStateBackend`.

So `isStateImmutableInStateBackend` can simply return `keyedStateBackend.requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT)`;

curcur replied:
keyedStateBackend.requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT) == true

means it  `use rocksdb` && `uses Heap Timer`

What needs to return is `use rocksdb` && `NOT use Heap Timer`

They are not the same, and you can not simply return `keyedStateBackend.requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT);`

rkhachatryan replied:
`requiresLegacySynchronousTimerSnapshots` is overriden in `RocksDBKeyedStateBackend` so there is no need to check the class.
Or am I missing something?

curcur replied:
The returned logic of isStateImmutableInStateBackend() is not the same as what requiresLegacySynchronousTimerSnapshots() returns, please check my highlights of the last comment.

requiresLegacySynchronousTimerSnapshots() == false only indicates it does not uses a heap timer; it does not indicate it is a rocksdb statebackend.

the required logic needs requiresLegacySynchronousTimerSnapshots() == false && uses rocks db

rkhachatryan replied:
Maybe it's easier to discuss having a code change.
I published a hotfix PR: https://github.com/apache/flink/pull/15046

curcur replied:
That logically makes more sense.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/delegate/DelegateKeyedStateBackend.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @param <K> Type of the key by which states are keyed.
+ */
+public interface DelegateKeyedStateBackend<K>
+        extends CheckpointableKeyedStateBackend<K>, CheckpointListener {
+
+    AbstractKeyedStateBackend<K> getDelegatedKeyedStateBackend();
```
1. As `StateConfigUtil.isStateImmutableInStateBackend` doesn't need to explicitly access delegated backend, this interface becomes unnecessary.
2. If not, I think it should be `@Experimental`
3. And `getDelegatedKeyedStateBackend` should return `KeyedStateBackend`

curcur replied:
1. As the reason mentioned above, it still needs to access delegated backend
2. Why need to add `@Experimental`? Do we want to remove it later? Do we need to add `@Experimental` in other places?
3. And getDelegatedKeyedStateBackend should return KeyedStateBackend (OK)

rkhachatryan replied:
So the interface is removed, right?

curcur replied:
Right.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/delegate/DelegateStateBackend.java:
```diff
@@ -0,0 +1,35 @@
+ * <p>TODO: provide configurable way to verify whether a state backend can be delegated.
+ */
+@Internal
+public interface DelegateStateBackend extends StateBackend {
```
`DelegateStateBackend` -> `DelegatingStateBackend`?
Just "Delegate" is ambiguous to me.

curcur replied:
For an "object", I'd prefer not to use "ing" format; it causes more confusion.

rkhachatryan replied:
I don't see any confusion with "ing".

But putting this aside, 
using this suffix is quite common in general, in Flink and even in this PR: `ChangelogReducingState`, `ChangelogAggregatingState`.

So I'd name it `DelegatingStateBackend` at least for consistency.

curcur replied:
They are different. 
First,
XXX Reducing State, reducing is quite obviously an adj-noun;

while

DelegatingStateBackend, I can not tell it is an adj-noun or action;

Second,
I feel "Reducing" and "Aggregating" themselves are confusing as well; For Map state, should we call it Mapping State?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedStateBackend.java:
```diff
@@ -136,6 +138,15 @@
      */
     boolean deregisterKeySelectionListener(KeySelectionListener<K> listener);
 
+    /** Returns the total number of state entries across all keys/namespaces. */
+    @VisibleForTesting
+    int numKeyValueStateEntries();
```
I'd like to avoid adding methods exclusively for testing to public production interfaces.
Instead, it can be in `AbstractKeyedStateBackend`.

curcur replied:
That's a very good point, thanks!

curcur replied:
I remembered I moved it from `AbstractKeyedStateBackend` to `KeyedStateBackend.java`.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -249,6 +266,66 @@ public static StateBackend fromApplicationOrConfigOrDefault(
+                        Class.forName(CHANGELOG_STATE_BACKEND, false, classLoader)
+                                .asSubclass(DelegateStateBackend.class)
+                                .getConstructor(StateBackend.class);
+                return constructor.newInstance(backend);
```
Should we check for double delegation here?

curcur replied:
No
`ChnagelogStateBackend` can only be loaded from `loadStateBackend`;
`loadStateBackend` loads the wrapped statebackend from `fromApplicationOrConfigOrDefault`
`fromApplicationOrConfigOrDefault` can only return non-delegated statebackend

rkhachatryan replied:
1. It's better to not rely on distant checks (as this method can potentially receive an argument from somewhere else; or the distant method can change)
2. I think the check in `loadStateBackend` should be removed.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogAggregatingState.java:
```diff
@@ -0,0 +1,129 @@
+ * @param <ACC> The type of the value stored in the state (the accumulator type).
+ * @param <OUT> The type of the value returned from the state.
+ */
+public class ChangelogAggregatingState<K, N, IN, ACC, OUT>
```
Looks like it can be package-private.

ditto other Changelog State classes

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -0,0 +1,342 @@
+ *
+ * @param <K> The key by which state is keyed.
+ */
+public class ChangelogKeyedStateBackend<K> implements DelegateKeyedStateBackend<K> {
```
package-private and `@Internal`?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -0,0 +1,342 @@
+        return keyedStateBackend.deregisterKeySelectionListener(listener);
+    }
+
+    @Override
+    public <N, S extends State, T> void applyToAllKeys(
```
I think this method can be refactored to avoid code duplication by:
1. Remove `AbstractKeyedStateBackend.supportConcurrentModification()`
2. Add an abstract method to iterate over keys to `AbstractKeyedStateBackend` (and override in heap backend)
3. Add `AbstractKeyedStateBackend.applyToAllKeys(Function<..., S> stateFactory, ...)` method
4. By default, use `AbstractKeyedStateBackend.getPartitionedState()`
5. In this class override it and by pass `this::getPartitionedState`
So here it would look like:
```
public void applyToAllKeys(...) {
    keyedStateBackend.applyToAllKeys(namespace, namespaceSerializer, stateDescriptor, function,
           this::getPartitionedState);
```

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    // -------------------- DelegateKeyedStateBackend --------------------------
+    public AbstractKeyedStateBackend<K> getDelegatedKeyedStateBackend() {
+        return keyedStateBackend;
```
This method is unnecessary anymore if `isStateImmutableInStateBackend` is simplified as commented above.
If not, please add `@Override` annotation.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:
```diff
@@ -0,0 +1,108 @@
+            @Nonnull Collection<KeyedStateHandle> stateHandles,
+            CloseableRegistry cancelStreamRegistry)
+            throws Exception {
+        LOG.debug("ChangelogStateBackend creates its KeyedStateBackend.");
```
I think this logging is unnecessary (ditto othere similar log statements).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateHashMapTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Test
+    public void testDelegatedValueState() throws Exception {
+        assertValueState(createKeyedBackend(IntSerializer.INSTANCE), HeapValueState.class);
```
I see some issues with ChangelogDelegate.*Tests:
1. Production code is made public (e.g. `HeapValueState`, `RocksDBValueState`)
2. Explicit assertions about each delegated backend implementation details (so changing RocksDb state class will break the tests)
3. A lot of code duplication

Why not have just one test class that would assert for every state type the returned delegated state
is the same as delegating:

```
StateBackend delegatedBackend = new HashMapStateBackend(); // implementation doesn't matter
State delegatedState = delegatedBackend.createKeyedStateBackend(...).getPartitionedState(...);
ChangelogStateBackend changelogBackend = new ChangelogStateBackend(delegatedBackend);
ChangelogValueState changeLogState = changelogBackend.createKeyedStateBackend().getPartitionedState();
assertEquals(delegatedState.getClass(), changeLogState.getDelegatedValueState().getClass());
```

(while still having almost empty sublasses to run `StateBackendTestBase` again different backend combinations)

curcur replied:
My original concern is to make sure when "StateBackendTestBase" is tested against ChnagelogStateBackend, we make sure the wrapper state is used instead of the wrapped state.

But I think one thing I really do not like is the creation and disposal of the state backend is not in the same method (to avoid duplicated code). So I agree to exact these test out as a new test and refactor it.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapValueState.java:
```diff
@@ -31,7 +31,7 @@
  * @param <V> The type of the value.
  */
-class HeapValueState<K, N, V> extends AbstractHeapState<K, N, V>
+public class HeapValueState<K, N, V> extends AbstractHeapState<K, N, V>
```
Please revert the visibility if you adjust tests as I suggested.

ditto: other state classes

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/pom.xml:
```diff
@@ -0,0 +1,91 @@
+
+        <dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-test-utils-junit</artifactId>
```
test scope?

rkhachatryan reviewed (changes requested):
Thanks for updating your PR @curcur.
I still have some concerns, please take a look at the comments below.

curcur commented:
@rkhachatryan Hey Roman, I believe I've addressed most of your comments except two:

1. "Flag in API" we discussed yesterday
2. Introduce test interface for `numKeyValueStateEntries`

For 1. I was thinking about whether we can put the flag in API in JM-config and TM-config; and initiate the wrapper loadStateBackend; the same place as:

`if (config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG)) {`

so `fromApplication` would still be kept as an unwrapped state bakcend.

====
You also mentioned that statebackend is loaded for recovery: do you mean something here?
Checkpoints#loadCheckpointStorage
Checkpoints#loadStateBackendFromConfig
StateBackendLoader#loadStateBackendFromConfig


For 2, I am still feeling introduce a test interface for `numKeyValueStateEntries` is a bit overkill...

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:
```diff
@@ -0,0 +1,105 @@
+ * This state backend holds the working state in the underlying delegatedStateBackend, and forwards
+ * state changes to State Changelog.
+ */
+public class ChangelogStateBackend implements DelegateStateBackend {
```
I think `ChangelogStateBackend` should also implement `ConfigurableStateBackend` and delegate `configure` call if needed.
Alteratively, it can be done in the `DelegateStateBackend` interface (default method).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -249,6 +266,66 @@ public static StateBackend fromApplicationOrConfigOrDefault(
+        final StateBackend backend =
+                fromApplicationOrConfigOrDefault(fromApplication, config, classLoader, logger);
+
+        if (config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG)) {
```
Is a similar branch missing in `loadStateBackendFromConfig`?
It is called for example from `StreamExecutionEnvironment.loadStateBackend` which I think should get a wrapped version.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -0,0 +1,331 @@
+        return keyedStateBackend.numKeyValueStateEntries();
+    }
+
+    @Override
+    public boolean isStateImmutableInStateBackend(CheckpointType checkpointOptions) {
+        return keyedStateBackend.requiresLegacySynchronousTimerSnapshots(checkpointOptions);
+    }
```
good catch

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedStateBackendForTest.java:
```diff
@@ -0,0 +1,29 @@
+
+/** A keyed state backend interface for internal testing purpose. */
+@Internal
+public interface KeyedStateBackendForTest {
```
How about `TestableKeyedStateBackend`?

And marking the interface itself `@VisibleForTesting`?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -173,6 +186,22 @@ public static StateBackend loadStateBackendFromConfig(
+
+        final StateBackend backend = loadStateBackendFromConfig(config, classLoader, logger);
+
+        if (config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG)) {
+            return loadChangelogStateBackend(backend, classLoader);
```
I think the check whether `backend` is already delegating is missing here (`loadStateBackendFromConfig` can construct one using factory class).

curcur replied:
Do you mean `loadStateBackendFromConfig` would return a StateChangeLogStateBackend?

For now, all the wrapping logic is to construct a `StateChangeLogStateBackend` is through 
`loadChangelogStateBackend`

If we want to load through factory class (which we do not for now), we can add load through the factory 
through  `loadChangelogStateBackend` as well.

I think that is logically simpler and cleaner.

rkhachatryan replied:
> Do you mean loadStateBackendFromConfig would return a StateChangeLogStateBackend?

Yes.

curcur replied:
How do you think about my proposal above? If you do not like my proposal, maybe we can discuss this after I push the changed code.

curcur replied:
sync-up offline, will add a pre-check instead.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -173,6 +186,22 @@ public static StateBackend loadStateBackendFromConfig(
         }
     }
 
+    public static StateBackend loadWrappedStateBackendFromConfig(
+            ReadableConfig config, ClassLoader classLoader, @Nullable Logger logger)
+            throws IllegalConfigurationException, DynamicCodeLoadingException, IOException {
```
This changes adds two new public methods, increasing the total number of methods to get a backend to four. Ideally, I think there should be just one public method.

I see that two existing methods are now marked with `@VisibleForTesting` but that's not clear enough I think.

Instead, why not
1. Make existing methods private and add some suffix like `private loadStateBackendFromConfig*Internal`
2. Add two corresponding methods (`loadStateBackendFromConfig` and `fromApplicationOrConfigOrDefault`) which would call original ones and wrap if needed
3. No changes to callers required
?

curcur replied:
Yes, that's a good idea. But there might be some more clean-up tests, but I like this idea better.

curcur replied:
This is done.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java:
```diff
@@ -0,0 +1,274 @@
+
+    private final String backendKey = CheckpointingOptions.STATE_BACKEND.key();
+
+    @Test
+    public void testLoadingDefault() throws Exception {
```
I think this is actually a test for `StateBackendLoader` so I'd prefer to move it to `StateBackendLoadingTest`.

Testing of RocksDb wrapping won't work, but I don't think we need to test for all combinations.

curcur replied:
The problem is we want to test loading ChangelogStateBackend. If I put it to run-time, it would have a runtime dependency problem

rkhachatryan replied:
Yes, you are right.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java:
```diff
@@ -0,0 +1,274 @@
+                backend, HashMapStateBackend.class, storage, JobManagerCheckpointStorage.class);
+    }
+
+    @Test
+    public void testLoadingRocksDBStateBackendFromConfig() throws Exception {
+        final Configuration config = config("rocksdb");
+        final StateBackend backend = loadStateBackend(config);
+        final CheckpointStorage storage =
+                CheckpointStorageLoader.load(null, null, backend, config(), cl, null);
+
+        assertDelegateStateBackend(
+                backend,
+                EmbeddedRocksDBStateBackend.class,
+                storage,
+                JobManagerCheckpointStorage.class);
+    }
```
This code is duplicated many times in this class.
I think duplication can be eliminated by extracting test parameters (backend name, backend class, storage class).
But as I said above, I don't think we need to test all these combinations.

curcur replied:
The main reason I've added all types of backend loading tests is because of the separation of checkpoint storage and statebackend (and all the default settings).

I want to make sure all types are correct. If later we deprecated and simplify rocksdb/fs/mem statebackend, I think these tests can be removed.

But I will try to deduplicate this.

curcur commented:
rebase and address Roman's comments.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/resources/log4j2.properties:
```diff
@@ -0,0 +1,28 @@
+# limitations under the License.
+################################################################################
+
+# Set root logger level to OFF to not flood build logs
+# set manually to INFO for debugging purposes
+rootLogger.level = OFF
+rootLogger.appenderRef.test.ref = TestLogger
```
I think we should only have `log4j2.properties` in `src/test/resources`.

curcur replied:
good catch

rkhachatryan reviewed (approved):
Thanks for updating your PR @curcur .
LGTM (except that `src/main/resources/log4j2.properties` is not needed - please remove it).

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java:
```diff
@@ -0,0 +1,109 @@
+ * @param <N> The type of the namespace.
+ * @param <V> The type of the value.
+ */
+class ChangelogValueState<K, N, V> implements InternalValueState<K, N, V> {
```
We could introduce `AbstractChangelogState` to act as `InternalKvState`, so that we could only keep only one place of those `getKeySerializer`, `getNamespaceSerializer`, `getSerializedValue` and so on.

Maybe you could refer to [pull/15119](https://github.com/apache/flink/pull/15119/files#diff-52de1f1a9a4e344e62306261f83533bc6a87326f372d775bc984c3a944d66943) for similar implementation.

curcur replied:
Thanks, I've tried this way, and I like it a lot better.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java:
```diff
@@ -0,0 +1,109 @@
+        valueState.clear();
+    }
+
+    public InternalValueState<K, N, V> getDelegatedValueState() {
```
Actually, I think all such `getDelegatedValueState` could be simplified to `getDelegatedState`, and we always keep just protected `delegatedState` within the `AbstractChangelogState`.

curcur replied:
Same as above.

Myasuka commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -193,24 +193,24 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
 
-    protected <K> AbstractKeyedStateBackend<K> createKeyedBackend(TypeSerializer<K> keySerializer)
-            throws Exception {
+    protected <K> CheckpointableKeyedStateBackend<K> createKeyedBackend(
```
Actually, I don't know why we have to change the returned class from `AbstractKeyedStateBackend` to `CheckpointableKeyedStateBackend`, is it necessary?

curcur replied:
The reason is that all the tests (including 4 different extensions of StateBackendTestBase.java) are reused by ChangelogStateBackend. 

"ChangelogKeyedStateBackend" is not an "AbstractKeyedStateBackend"; it is a `CheckpointableKeyedStateBackend` instead, that's why the change is made.

If not made this way, we would have to rewrite and duplicatete all the tests (which are similar to what are included in StateBackendTestBase)

Myasuka commented on flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java:
```diff
@@ -96,6 +96,16 @@
+    @Documentation.Section(value = Documentation.Sections.COMMON_STATE_BACKENDS)
+    @Documentation.ExcludeFromDocumentation("Hidden for now")
+    public static final ConfigOption<Boolean> ENABLE_STATE_CHANGE_LOG =
+            ConfigOptions.key("state.backend.enable-statechangelog")
```
I think this option targets for checkpointing, and maybe `state.checkpoints.enable-changelog` looks better.

rkhachatryan replied:
To me, this parameter changes the behavior of backend in how it performs checkpointing. So I'd leave it under `state.backend` section. 

But probably `state.backend.changelog.enabled` would be better as there will be more settings (and it got a bit more votes in an offline vote:) )

curcur replied:
I agreed with what Roman said,

the state changelog is related to how the state is maintained and handled, and as a whole changing the behavior of checkpoint as well.

"state.backend.changelog.enabled" yes, this is better.

Myasuka reviewed (commented):
I noticed that there existed many code duplication, and I also try to introduce latency tracking state which also wraps original state. I think we could remove such code duplication when wrapping changelog based state.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:
```diff
@@ -185,16 +186,21 @@ public static CheckpointStorage load(
-        // Legacy state backends always take precedence
-        // for backwards compatibility.
-        if (configuredStateBackend instanceof CheckpointStorage) {
+        // Legacy state backends always take precedence for backwards compatibility.
+        StateBackend rootStateBackend =
+                (configuredStateBackend instanceof DelegatingStateBackend)
+                        ? ((DelegatingStateBackend) configuredStateBackend)
+                                .getDelegatedStateBackend()
+                        : configuredStateBackend;
+
+        if (rootStateBackend instanceof CheckpointStorage) {
```
nit: wouldn't it be better to replace this with a default method in `StateBackend`:

```
interface StateBackend {
...
default boolean isCheckpointStorage() {
  return this instanceof CheckpointStorage;
}
...
}
```
and overload this method in `DelegatingStateBackend`/`ChangelogStateBackend`?

curcur replied:
Maybe we can separate this to a different PR? I think this should be a follow-up of 
"FLINK-19463 Disentangle StateBackends from Checkpointing"

and find all places of `if (rootStateBackend instanceof CheckpointStorage)`?

curcur commented:
@Myasuka , thanks for reviewing the PR. I like your suggestions on how to deduplicate the state access part, which in return also helps reducing a lot of duplication code on state tests as well. Thanks again.

Please take a look whether my change makes sense to you.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateStateTest.java:
```diff
@@ -0,0 +1,134 @@
+            throws Exception {
+        KeyedStateBackend<Integer> delegatedBackend = createKeyedBackend(backend.get(), env);
+        KeyedStateBackend<Integer> changelogBackend =
+                createKeyedBackend(new ChangelogStateBackend(backend.get()), env);
```
Why we have to create the `delegatedBackend` explicitly? Can we just get the delegated keyed backend from `changelogBackend` directly?

curcur replied:
That's what we want to compare against.

One is created from the unwrapped state backend, and the other is created from the wrapped state backend to make sure the wrapped one has the right wrapped state + the underlying state class is consistent with the unwrapped one.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateStateTest.java:
```diff
@@ -0,0 +1,134 @@
+                                    descriptor)
+                            .getClass());
+        } finally {
+            delegatedBackend.dispose();
```
We'd better add not null check and then disposing those keyed state backends.

curcur replied:
OK

Myasuka reviewed (commented):
Thanks for your update, please take a look at my new comments.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -0,0 +1,336 @@
+    // -------- Methods not simply delegating to wrapped state backend ---------
+    @Override
+    @SuppressWarnings("unchecked")
+    public <N, S extends State, T> S getOrCreateKeyedState(
```
Here `keyedStateBackend` you mean the delegated state backend right?
Wouldn't `keyedStateBackend` (delegated keyedstatebackend) use its own factory to create its own state (unwrapped) and put in its own cache?

And that's the main reason why we want to delegate state and change the creation/access?

Myasuka replied:
I think you're right, we have to make the two methods looking some kind of code duplication from `AbstractKeyedStateBackend`.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -0,0 +1,336 @@
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <N, S extends State> S getPartitionedState(
```
Same as above.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -249,6 +306,46 @@ public static StateBackend fromApplicationOrConfigOrDefault(
+            return loadChangelogStateBackend(backend, classLoader);
+        } else {
+            LOG.info(
+                    "Delegate State Backend is not used, and the State Backend is {}",
```
Yep, the logic has been changed upon its first version. Good catch!

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Override
+    public <K> ChangelogKeyedStateBackend<K> createKeyedStateBackend(
```
Not having the full context, I would be a bit hesitant to do so, because the suggested change is made based on the assumption that the default `managedMemoryFraction` to createKeyedStateBackend is always `1.0`. 
I would prevent hard-coding the default value as such.

Or put another way, why we have two different methods in the interface `StateBackend`

`createKeyedStateBackend`, and why we do not simply hardcode 1.0 in one of the methods in a similar way as you suggested?

I would be cautious to make the change for a safety purpose.

Myasuka replied:
The `1.0` should not change in a long term, but I see, we could prefer to safety here.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractChangelogState.java:
```diff
@@ -0,0 +1,83 @@
+    public S getDelegatedState() {
+        return delegatedState;
+    }
+
```
I was thinking of this when extracting the abstract class.

But I am not sure whether we would have extra/different functionalities when clearing different types of states. I only extract the methods I am sure they are the same.

carp84 commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StateConfigUtil.java:
```diff
@@ -46,10 +41,7 @@ public static StateTtlConfig createTtlConfig(long retentionTime) {
-                && ((AbstractKeyedStateBackend<?>) stateBackend)
-                        .isStateImmutableInStateBackend(CheckpointType.CHECKPOINT);
+    public static boolean isStateImmutableInStateBackend(KeyedStateBackend<?> keyedStateBackend) {
+        return keyedStateBackend.isStateImmutableInStateBackend(CheckpointType.CHECKPOINT);
```
```suggestion
        // TODO: remove the hard code check once FLINK-21027 is supported
        return keyedStateBackend.isStateImmutableInStateBackend(CheckpointType.CHECKPOINT);
```
We need to keep this TODO comment since invocation of this `isStateImmutableInStateBackend` method should be replaced by the to-be-added `isStateKeyValueSerialized` method after FLINK-21027 is implemented.

carp84 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimeServiceManagerImpl.java:
```diff
@@ -93,17 +93,17 @@ private InternalTimeServiceManagerImpl(
      */
     public static <K> InternalTimeServiceManagerImpl<K> create(
-            CheckpointableKeyedStateBackend<K> keyedStatedBackend,
+            CheckpointableKeyedStateBackend<K> keyedStateBackend,
```
I could find many nice cleanups like this but a little bit distracted for review. Maybe a separate commit for such cleanups next time (smile)

carp84 commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:
```diff
@@ -0,0 +1,163 @@
+        this.delegatedStateBackend = Preconditions.checkNotNull(stateBackend);
+
+        Preconditions.checkArgument(
+                !(stateBackend instanceof ChangelogStateBackend),
```
```suggestion
                !(stateBackend instanceof DelegatingStateBackend),
```

carp84 reviewed (changes requested):
Overall LGTM, just some minor suggestions. Thanks for the efforts @curcur .

curcur commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/mock/MockKeyedStateBackend.java:
```diff
@@ -182,12 +181,6 @@ public int numKeyValueStateEntries() {
-        return false;
-    }
-
-    @Override
```
You are right, @Override should not be removed

the removed method is not necessary. I probably should have another PR to do this next time.

curcur commented:
@carp84 Thanks for your time! I've addressed the comments, please take a look

carp84 reviewed (approved):
Thanks for the quick response @curcur , the latest changes LGTM.

curcur commented:
rebase
